base/file.jl

```diff
@@ -513,20 +513,37 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
     return (filename, Base.open(filename, "r+"))
 end
 
+# GUID  win32 api (consider # TODO: refactor)
+struct GUID
+    l1::Culong
+    w1::Cushort
+    w2::Cushort
+    b::NTuple{8,Cuchar}
+end
+
 function tempname(parent::AbstractString=tempdir(); cleanup::Bool=true)
     isdir(parent) || throw(ArgumentError("$(repr(parent)) is not a directory"))
-    seed::UInt32 = rand(UInt32)
-    while true
-        if (seed & typemax(UInt16)) == 0
-            seed += 1
-        end
-        filename = _win_tempname(parent, seed)
-        if !ispath(filename)
-            cleanup && temp_cleanup_later(filename)
-            return filename
-        end
-        seed += 1
-    end
+
+    RPC_S_UUID_LOCAL_ONLY = 1824
+    RPC_S_UUID_NO_ADDRESS = 1739
+    RPC_S_OUT_OF_MEMORY = 14
+
+    id = Ref{GUID}()
+    r = ccall((:UuidCreate,:Rpcrt4), stdcall, Cint, (Ref{GUID},), id)
+    (r == 0 || r == RPC_S_UUID_LOCAL_ONLY) || (r == RPC_S_UUID_NO_ADDRESS && error("Cannot get Ethernet or token-ring hardware address for this computer."))
+
+    nameptr = Ref{Ptr{Cwchar_t}}()
+    r = ccall((:UuidToStringW, :Rpcrt4), stdcall, Cint, (Ref{GUID}, Ref{Ptr{Cwchar_t}}), id, nameptr)
+    r == 0 || (r == RPC_S_OUT_OF_MEMORY && error("Out of memory")) || error("cannot convert UIUD to string")
+
+    namebuf = unsafe_wrap(Vector{Cwchar_t}, nameptr[], ccall(:wcslen, UInt, (Ptr{Cwchar_t},), nameptr[]))
+    name = transcode(String, namebuf)
+
+    ccall((:RpcStringFreeW, :Rpcrt4), stdcall, Cint, (Ref{Ptr{Cwchar_t}},), nameptr) == 0
+
+    filename = joinpath(parent, temp_prefix * name)
```

Comment:
I'm tempted to say we should also do:
```suggestion
    @assert !ispath(filename)
```
to guard against some extreme forms of malfeasance.

Thoughts?

Reply:
I'm not sure if we want to do this. I prefer not to include this since, ideally, this function should avoid touching the filesystem at all.

Reply:
It's expected to touch the filesystem in this way on all other platforms though (it's part of the definition of posix tempnam we use otherwise)

Reply:
What if we just restrict this to the branch where `cleanup==true` ?